### PR TITLE
WIP: Add GetOrReserveArtifact API

### PR DESCRIPTION
**Pull request by milton0825**
# TL;DR
Create an API `GetOrReserveArtifact` that will be used by flytepropeller to reserve a spot for populating the artifact if the artifact does not exist yet.

Will implement `ExtendReservation` & `ReleaseReservation` APIs in separate PRs.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Create an API `GetOrReserveArtifact` that will be used by `flytepropeller` to reserve a spot for populating the artifact if the artifact does not exist yet. A new `Reservation` table is created to track which task is occupying the spot.

Check https://docs.google.com/document/d/1Oq8f0yVoHMHDCUbxhu9V-4dDLzbbM4YWcnNNscEnYkM/edit#heading=h.ipjr4byywfgm for more details,

## Tracking Issue
https://github.com/flyteorg/flyte/issues/872

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: pkg/manager/impl/reservation_manager.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		now:                nowFunc,
+		systemMetrics:      systemMetrics,
+	}
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
```
Comment (milton0825) [inline pkg/manager/impl/reservation_manager.go:85]: change errors2..

**Comment by milton0825** [comment]
cc: @kumare3 @EngHabu

File: pkg/manager/impl/reservation_manager.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+			// Reservation does not exist yet so let's create one
+			err := repo.Create(ctx, models.Reservation{
+				ReservationKey: reservationKey,
+				OwnerID:        request.OwnerId,
+				ExpireAt:       r.now().Add(r.reservationTimeout),
+			})
+
+			if err != nil {
```
Comment (katrogan) [inline pkg/manager/impl/reservation_manager.go:131]: what about the race condition case where two processes attempt to create the same reservation after asserting it doesn't exist? will the primary key guarantee a failure due to duplication for the second call?

File: pkg/manager/impl/reservation_manager.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	if rsv.ExpireAt.Before(now) {
+		// The reservation is expired, let's try to grab the reservation
+		rowsAffected, err := repo.Update(ctx, reservationKey,
+			rsv.ExpireAt,
+			now.Add(r.reservationTimeout), request.OwnerId)
+		if err != nil {
+			return datacatalog.ReservationStatus{}, err
+		}
+
+		// If rows affected is 0, possibly someone else is also
```
Comment (katrogan) [inline pkg/manager/impl/reservation_manager.go:159]: if rowAffected still 0 when the owner id is different?

File: pkg/manager/interfaces/reservation.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package interfaces
+
+import (
+	"context"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/datacatalog"
+)
+
+type ReservationManager interface {
```
Comment (katrogan) [inline pkg/manager/interfaces/reservation.go:9]: Mind adding a comment describing what this public interface does?

File: pkg/repositories/gormimpl/reservation.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+func (r *reservationRepo) Get(ctx context.Context, reservationKey models.ReservationKey) (models.Reservation, error) {
+	timer := r.repoMetrics.GetDuration.Start(ctx)
+	defer timer.Stop()
+
+	var reservation models.Reservation
+
+	result := r.db.Where(&models.Reservation{
+		ReservationKey: reservationKey,
+	}).First(&reservation)
```
Comment (katrogan) [inline pkg/repositories/gormimpl/reservation.go:48]: nit: we've observed better performance using [Take](https://gorm.io/docs/query.html#Retrieving-a-single-object) rather than First since the former avoids an unnecessary sort

File: pkg/repositories/interfaces/reservation_repo.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package interfaces
+
+import (
+	"context"
+	"time"
+
+	"github.com/flyteorg/datacatalog/pkg/repositories/models"
+)
+
+type ReservationRepo interface {
```
Comment (katrogan) [inline pkg/repositories/interfaces/reservation_repo.go:10]: Public comment here too please

File: pkg/repositories/models/reservation.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package models
+
+import "time"
+
+type ReservationKey struct {
```
Comment (katrogan) [inline pkg/repositories/models/reservation.go:5]: comments here as well

**Review by katrogan — COMMENTED** [review]
Is the eventual goal that whoever makes the call to create or update an artifact must already own a reservation?

File: pkg/repositories/gormimpl/reservation.go (hunk 5 above)
Comment (kumare3) [reply pkg/repositories/gormimpl/reservation.go:48]: I dont think this is a nit, this is a requirement, because the api is called from flytepropeller

File: pkg/manager/impl/reservation_manager.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Tag does not exist yet, let's reserve a spot to work on
+			// generating the artifact.
+			status, err := r.makeReservation(ctx, request)
```
Comment (kumare3) [inline pkg/manager/impl/reservation_manager.go:88]: what happens if reservations fails with already exists.

Ideally the `Get` or `Reserve` should be one API and will need a transactional Get.

What you have done is fine too, as long as you have an already exists handling right

File: pkg/manager/impl/reservation_manager.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+				OwnerId: request.OwnerId,
+			}, nil
+		}
+		return datacatalog.ReservationStatus{}, err
+	}
+
+	now := r.now()
+	// Reservation already exists so there is a task already working on it
+	// Let's check if the reservation is expired.
+	if rsv.ExpireAt.Before(now) {
```
Comment (kumare3) [inline pkg/manager/impl/reservation_manager.go:150]: this is weird too. I think we should make this entire thing one single request

GetOrMake(pass the time)
// in db psuedo code
```python
if record is none or record is expired:
   create new record
   return
return record
```

File: pkg/repositories/gormimpl/reservation.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+	if result.Error != nil {
+		return reservation, r.errorTransformer.ToDataCatalogError(result.Error)
+	}
+
+	return reservation, nil
+}
+
+func (r *reservationRepo) Update(ctx context.Context, reservationKey models.ReservationKey, prevExpireAt time.Time, expireAt time.Time, ownerID string) (int64, error) {
```
Comment (kumare3) [inline pkg/repositories/gormimpl/reservation.go:57]: you need a conditional upsert

**Review by kumare3 — COMMENTED** [review]
Hi @milton0825 Thank you for your PR.
Few problems, Because of the nature of the API, it is necessary to use conditional writes for creating the reservation, else you will end up in unnecessary failures.

For performance reasons it might be necessary to perform a Get First, but
ideally you should do a conditional upsert

File: pkg/manager/impl/reservation_manager.go (hunk 8 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:88]: If the reservation already exists, on https://github.com/flyteorg/datacatalog/blob/c57349c955b680ac8cbc1ce173829e23a62971cd/pkg/manager/impl/reservation_manager.go#L150 we check if the existing reservation has expired and try to grab the reservation via conditional updatee.

File: pkg/manager/impl/reservation_manager.go (hunk 9 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:150]: So are you propose is to merge all the GET reservation, Create reservation and Update reservation db calls into a single Upsert request and do a GET in the end to retrieve the owner ID (necessary because we want to know who actually owns the reservation)

The reason for the separate API approach is that a GET RESERVATION call is needed to retrieve the owner of the reservation (possibly other task owns the reservation). Since we are doing a GET at first, why not just do the expiration time check at application layer so that we can avoid unnecessary db calls.

File: pkg/manager/impl/reservation_manager.go (hunk 2 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:131]: Yeah the primary key will ensure the second attempt fails with already exists.

**Comment by milton0825** [comment]
> Is the eventual goal that whoever makes the call to create or update an artifact must already own a reservation?

Correct.

File: pkg/manager/impl/reservation_manager.go (hunk 9 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:150]: The reason for the separate API approach is that GET RESERVATION call is always needed to retrieve the owner of the reservation . Since we are doing a GET at first, why not just do the expiration time check at application layer so that we can avoid unnecessary db calls.

File: pkg/manager/interfaces/reservation.go (hunk 4 above)
Comment (milton0825) [reply pkg/manager/interfaces/reservation.go:9]: Will add docs once we are finalized on the interface..

**Comment by milton0825** [comment]
> Few problems, Because of the nature of the API, it is necessary to use conditional writes for creating the reservation, else you will end up in unnecessary failures.

What unnecessary failures are you referring to here?

File: pkg/manager/impl/reservation_manager.go (hunk 9 above)
Comment (kumare3) [reply pkg/manager/impl/reservation_manager.go:150]: I agree that the get can be optimized, the only thing is if Get fails, you could still end up in a race. So, inside the i failure, we should do a conditional write and get or a Compare-and-swap operation. If we cannot do that, then maybe just do the upsert part atleast

Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:150]: > I agree that the get can be optimized, the only thing is if Get fails, you could still end up in a race. So, inside the i failure, we should do a conditional write and get or a Compare-and-swap operation

Can you elaborate a bit on this?

**Comment by kumare3** [comment]
> > your right will fail because even if you checked that the reservation has expired, when you are about to write, the original process may heartbeat and extend the reservation, as in the current code, they are modeled as discrete calls.
> 
> Yeah the first process would do a conditional write on the `owner_id` and update the `expired_at` when heartbeating and the second process would do a conditional write on the `expired_at` and update the `owner_id`. Only one process will win and the other process will fail.
> 
> Process 1:
> 
> ```
> UPDATE reservation SET expire_at = new date
> WHERE reservation_key = ... AND owner_id = 'process 1'
> ```
> 
> Process 2:
> 
> ```
> UPDATE reservation SET expire_AT = new date AND owner_id = 'process 2'
> WHERE reservation_key = ... AND expire_at = prev date
> ```

cool, this is a conditional update :) and insert too right so upsert, because it may happen that the reservation may not exist but get created after the get check

**Comment by kumare3** [comment]
@milton0825 are you still working on this?

**Comment by milton0825** [comment]
Yeah this was previously blocked by the gorm upgrade because we need the upsert feature.

File: pkg/repositories/models/reservation.go (hunk 7 above)
Comment (milton0825) [reply pkg/repositories/models/reservation.go:5]: done

File: pkg/repositories/interfaces/reservation_repo.go (hunk 6 above)
Comment (milton0825) [reply pkg/repositories/interfaces/reservation_repo.go:11]: added

File: pkg/repositories/gormimpl/reservation.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		clause.OnConflict{
+			Where:     clause.Where{Exprs: expressions},
+			UpdateAll: true,
+		},
+	).Create(&reservation)
+	if result.Error != nil {
+		return r.errorTransformer.ToDataCatalogError(result.Error)
+	}
+
+	if result.RowsAffected == 0 {
```
Comment (katrogan) [inline pkg/repositories/gormimpl/reservation.go:67]: interesting, so if the conditional clause isn't met there won't be any error returned?

File: pkg/repositories/models/reservation.go (hunk 7 above)
Comment (katrogan) [inline pkg/repositories/models/reservation.go:5]: can you add comments to these public structs?

File: pkg/repositories/transformers/reservation.go [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package transformers
+
+import (
+	"github.com/flyteorg/datacatalog/pkg/repositories/models"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/datacatalog"
+)
+
+func ToReservationKey(datasetID datacatalog.DatasetID, tagName string) models.ReservationKey {
```
Comment (katrogan) [inline pkg/repositories/transformers/reservation.go:8]: add a unit test?

File: pkg/repositories/models/reservation.go [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+	DatasetProject string `gorm:"primary_key"`
```
Comment (katrogan) [inline pkg/repositories/models/reservation.go:6]: can you delete a dataset or tag? should this be relational?

**Review by katrogan — COMMENTED** [review]
just to double check, were you able to test this locally in your sandbox?

**Comment by milton0825** [comment]
@katrogan no I haven't. Just want to make sure that we are aligned with this approach before I invest more effort to add more tests & integration testing on sandbox

**Comment by katrogan** [comment]
@milton0825 got it 👍 

@EngHabu still want to take a look?

File: pkg/manager/impl/reservation_manager.go [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+		Value: &datacatalog.GetOrReserveArtifactResponse_Artifact{
+			Artifact: artifact,
+		},
+	}, nil
+}
+
+func (r *reservationManager) tryAcquireReservation(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (datacatalog.ReservationStatus, error) {
+	repo := r.repo.ReservationRepo()
+	reservationKey := transformers.ToReservationKey(*request.DatasetId, request.TagName)
+	rsv, err := repo.Get(ctx, reservationKey)
```
Comment (EngHabu) [inline pkg/manager/impl/reservation_manager.go:116]: Why do we try to do a `Get` here? Why do we not always try to CreateOrUpdate? is it to optimize the write?

File: pkg/repositories/interfaces/reservation_repo.go [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+// Interface to interact with Reservation Table
+type ReservationRepo interface {
+
+	// Get reservation
+	Get(ctx context.Context, reservationKey models.ReservationKey) (models.Reservation, error)
+
+	// Create the reservation. If the reservation already exists, we try to take over the
+	// reservation via update.
```
Comment (EngHabu) [inline pkg/repositories/interfaces/reservation_repo.go:17]: Please add to the docs here a more detailed explanation of how that happens... and the role `expires_at` plays as well as the conditional write we do...

File: pkg/manager/impl/reservation_manager.go (hunk 14 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:116]: Yeah this is an optimization to reduce the number of writes to db. We always need to do a `GET` here because we want to know who owns the reservation and show it to users on the UI. However, the reservation is held by a single task most of the times and there is no need to do a write.

Comment (EngHabu) [reply pkg/manager/impl/reservation_manager.go:116]: awesome, please add some docs here to explain that..

File: pkg/manager/impl/reservation_manager.go [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+		if errors2.IsDoesNotExistError(err) {
+			// Reservation does not exist yet so let's create one
+			resvExists = false
+		} else {
+			return datacatalog.ReservationStatus{}, err
+		}
+	}
+
+	now := r.now()
+	if !resvExists || rsv.ExpireAt.Before(now) {
```
Comment (EngHabu) [inline pkg/manager/impl/reservation_manager.go:129]: We need some epsilon value to update... maybe it should be configurable (next to `ReservationTimeout`) or maybe it should just be a multiplier of that... so basically the condition here should look something like this:

```suggestion
	if !resvExists || rsv.ExpireAt.Before(now.Add(r.reservationTimeoutBuffer)) {
```

Or:
```suggestion
	if !resvExists || rsv.ExpireAt.Before(now.Add(r.reservationTimeout / r.reservationTimeoutGracePeriodMultiplier)) {
```

**Comment by milton0825** [comment]
Cool sounds like we are aligned with the approach and I will be adding docs/tests.

File: pkg/manager/impl/reservation_manager.go (hunk 16 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:129]: Can you explain a bit more on why we need `now.Add(r.reservationTimeout / r.reservationTimeoutGracePeriodMultiplier)`? We already store `now.Add(r.reservationTimeout)` to `expireAt` at the time we persist the reservation.

File: pkg/repositories/gormimpl/reservation.go (hunk 11 above)
Comment (milton0825) [reply pkg/repositories/gormimpl/reservation.go:67]: Correct. If no rows affected by the update, it will not return an error.

File: pkg/repositories/models/reservation.go (hunk 13 above)
Comment (milton0825) [reply pkg/repositories/models/reservation.go:6]: There is no datacatalog API for deleting a dataset/tag.

File: pkg/repositories/gormimpl/reservation.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+	"gorm.io/gorm/clause"
+)
+
+type reservationRepo struct {
+	db               *gorm.DB
+	repoMetrics      gormMetrics
+	errorTransformer errors2.ErrorTransformer
+}
+
+func NewReservationRepo(db *gorm.DB, errorTransformer errors2.ErrorTransformer, scope promutils.Scope) interfaces.ReservationRepo {
```
[the remark this answers is not in the document]
Comment (milton0825) [reply pkg/repositories/gormimpl/reservation.go:25]: Added. The docs for `Get` and `CreateOrUpdate` is in the interface.

File: pkg/manager/impl/reservation_manager.go (hunk 14 above)
Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:122]: done

File: pkg/repositories/interfaces/reservation_repo.go (hunk 15 above)
Comment (milton0825) [reply pkg/repositories/interfaces/reservation_repo.go]: done

File: pkg/manager/impl/reservation_manager.go (hunk 16 above)
Comment (EngHabu) [reply pkg/manager/impl/reservation_manager.go:135]: I realized what I'm looking for, it should be called `heartbeatInterval` and it should do `now.Add(r.heartbeatInterval * 2)` or something... here is why:

t0: client requests reservation
t0+heartbeat: client requests reservation (heartbeat) and is returned the same reservation
t0+reseravtionTimeout-heartbeat: client requests reservation (heartbeat) and is returned the same reservation (since it hasn't expired yet)
t0+reservationTimeout+epsilon (since clocks are not synced): a different client requests the reservation and wins because it now has expired... right? and this is bad...

If, on the other side, you add this `heartbeatInterval*2` check, what will happen is:
t0: client requests reservation
t0+heartbeat: client requests reservation (heartbeat) and is returned the same reservation
t0+reseravtionTimeout-heartbeat: client requests reservation (heartbeat) and is returned an 'extended' reservation (since it'll be considered expired/about to expire)
t0+reservationTimeout+epsilon (since clocks are not synced): a different client requests the reservation, it'll lose because the original client still has the reservation

Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:135]: Let me clarify a bit. The `reservation_timeout` is essentially a heartbeat timeout. Once acquired a reservation, the client is responsible to heartbeat by calling `ExtendReservation` API at an interval less than the `reservation_timeout` otherwise the reservation may expire and the other client can take over the reservation. Let's use an example to illustrate this and let's assume the `reservation_timeout` is 5 min.

| Time | What | expireAt |
| --- | --- | --- |
| 00:00 | A try to acquire reservation via GetOrReserveArtifact API and succeeded | 00:05 |
| 00:02 | A heartbeats via ExtendReservation and that reset expireAt to now (00:02)+reservation_timeout (5min) = 00:08  | 00:08 |
| 00:05 | B try to acquire reservation via GetOrReserveArtifact API and failed | 00:08 |
| 00:09 | C try to acquire reservation via GetOrReserveArtifact API and succeeded. A did not heartbeat again by 00:08 so reservation expired and C took over the reservation | -- |

Comment (milton0825) [reply pkg/manager/impl/reservation_manager.go:135]: And if we set `reservation_timeout` to be 2-3X heartbeat interval, we could avoid the issue caused by clock not sync'd issue.

File: pkg/manager/impl/reservation_manager.go [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+		}
+
+		r.systemMetrics.reservationAcquired.Inc(ctx)
+		return datacatalog.ReservationStatus{
+			State:   datacatalog.ReservationStatus_ACQUIRED,
+			OwnerId: request.OwnerId,
+		}, nil
+	}
+
+	logger.Debugf(ctx, "Reservation: %+v is hold by %s", reservationKey, rsv.OwnerID)
```
Comment (EngHabu) [inline pkg/manager/impl/reservation_manager.go:180]:
```suggestion
	logger.Debugf(ctx, "Reservation: %+v is held by %s", reservationKey, rsv.OwnerID)
```

File: pkg/manager/impl/reservation_manager.go (hunk 16 above)
Comment (EngHabu) [reply pkg/manager/impl/reservation_manager.go:135]: aha, so that's what I was missing... for some reason I assumed GetOrReserve is the only available api and hence where the extension should happen...

Alright, so let's explore a startup scenario (after a deployment or crash... etc.).. the client starts up and it should immediately try to `ExtendReservation`, right? there doesn't seem to be a value in calling `GetOrReserve` in that case because even if the reservation is returned, it might expire as soon as it's received by the client...
So the only use-case for GetOrReserve is when it's the very first time it's reserving (or if it knows it lost the reservation)... after that it should only rely on ExtendReservation until it gets a response saying it lost the reservation...

I was thinking that even for the startup scenario, the client will call GetOrReserve (and not have to maintain state) and can reliably trust the response here... the way the code is done now, this can happen:
```
resp := GetOrReserve()
if resp.OwnerId == me {
    err := ExtendReservation
    👎🏾  it fails because in between the two calls, it expired and somebody else stole it...
}
```

if you do add the heartbeat interval to the check, GetOrReserve will be much more stable in that regard, right?

Another question RE: ExtendReservation, will it fail if the reservation has indeed expired but hasn't been claimed by anybody else? I guess it should just succeed, right?
